[Inductor] Implement a deduplist data structure for name to user tracking

oulgen opened the pull request:
Summary:
An internal MRS model was taking over a day's worth of time to compile due to many duplicates in dependency tracking. This PR replaces the list with a custom dedup list.
Normally one could use a set/dict for this purpose however the list in question gets elements appended as it is being iterated over which means that we need to keep the list semantics.

Test Plan: ad hoc testing

Differential Revision: D52060659




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

aakhundov commented on torch/_inductor/scheduler.py:
```diff
@@ -1330,7 +1342,39 @@ def compute_dependencies(self):
+                self.items = items or list()
+                self.membership = membership or set()
+
+            def append(self, node_user: T) -> None:
```
The dedup logic added to the list is limited to the `append` only. Things like `insert`, `pop`, etc. won't be consistent with the deduping. Do we want to reflect this in the docstring?

oulgen replied:
I did not implement those methods and i am not extending the list implementation here, so these methods will just error

aakhundov commented on torch/_inductor/scheduler.py:
```diff
@@ -1330,7 +1342,39 @@ def compute_dependencies(self):
+            """
+            This data structure behaves like a list except it makes sure the
+            elements remain unique.
+            Normally one could use a set/dict for this purpose however
+            the list in question gets elements appended as it is being
+            iterated over which means that we need to keep the list
+            semantics.
```
So `dict` is unique and order-preserving. Can't we use `dict` for this use case (i.e., adding while we enumerate)?

oulgen replied:
Dict also do not allow for appending in the middle of iteration

aakhundov commented on torch/_inductor/scheduler.py:
```diff
@@ -1330,7 +1342,39 @@ def compute_dependencies(self):
+
+            def __add__(self, other: "DedupList[T]") -> "DedupList[T]":
+                new_membership = set.union(self.membership, other.membership)
+                # Potential duplicates but we can be a little bit more
+                # pessimistic here
+                new_items = self.items + other.items
```
Maybe use `(new_membership - set(self.items))` instead of `other.items`? Then we won't have to be pessimistic.

oulgen replied:
new_membership is not order preserving, so i cant actually use that for new_items

Actually this whole `__add__` implementation is silly since it only ever gets called on empty list if you look at the code below

aakhundov replied:
Maybe remove it and refactor the call site then?

mlazos commented on torch/_inductor/scheduler.py:
```diff
@@ -1405,7 +1449,7 @@ def add_user(used_by_name, user_node, can_inplace=False, is_weak=False):
                 add_user(alt_name, node)
                 node.add_mutation_dep(StarDep(alt_name))
-                for other_node in name_to_users[alt_name]:
+                for other_node in name_to_users[alt_name].items:
```
Can you explain why appending in the middle of an iteration is required here? From what I understand, we are iterating over the users of node A, and appending to their respective lists, and because node A is not dependent on itself, we should never be touching the list we are iterating over, so I would think a set should work here.

oulgen replied:
I tried using a dict first (since sets do not have deterministic iteration order) and got the error message "dict was modified during iteration". So it looks like `add_user` below somehow modifies name_to_users for the node we are iterating over

int3 replied:
Which test surfaced this error? Just curious

oulgen replied:
It is a custom kernel on D51815383

jansel commented on torch/_inductor/scheduler.py:
```diff
@@ -1330,7 +1342,39 @@ def compute_dependencies(self):
+                new_membership = set.union(self.membership, other.membership)
+                # Potential duplicates but we can be a little bit more
+                # pessimistic here
+                new_items = self.items + other.items
```
Either remove this or do something like:
```suggestion
                new_items = self.items + [x for x in other.items if x not in self.membership]
```

As implemented this could result in duplicates.

oulgen committed:
[Inductor] Implement a deduplist data structure for name to user tracking (#115609)

Summary:

An internal MRS model was taking over a day's worth of time to compile due to many duplicates in dependency tracking. This PR replaces the list with a custom dedup list.
Normally one could use a set/dict for this purpose however the list in question gets elements appended as it is being iterated over which means that we need to keep the list semantics.

Test Plan: ad hoc testing

Differential Revision: D52060659

oulgen commented:
@pytorchbot merge
